### DISCUSSION TITLE
Crash on attempt to write from i2S mic to SD card

### DISCUSSION THREAD

**Discussion by musiq021** [pschatzmann/arduino-audio-tools#985] [Q&A] [answered]
( Sketch in question : https://pastebin.com/6KPetyeq )

I'm sure it's something simple, a basic minsunderstanding of the order of operations / flow but after many hours hunting I'm out of ideas. I assume I'm corrupting memory somehow?

When I call startRec, it seems to succesfuly open the file for writing "Opened r1.wav to record. " but then immediately reboots.

Using Arduino IDE 2, a ESP32 S2 Mini, with an INMP441 mic and a FAT 32GB SD card.

Playing WAVs from SD card working. 
Microphone appears to be wired right ( sketch recording and playback from RAM works ).

**Comment by musiq021** [comment]
Actually, it prints out the following before rebooting : 
[I] CodecWAV.h : 482 - sample_rate: 16000
[I] CodecWAV.h : 483 - channels: 1
[I] CodecWAV.h : 492 - is_streamed! because length is 2147418112
[I] StreamCopy.h : 82 - buffer_size=1024

**Comment by pschatzmann** [comment, marked as answer]
https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working

Test the SD functionality outside of the AudioTools to make sure that this is working properly!

**Reply by musiq021** [reply]
The test sketch works fine, but I just noticed it generates a bunch of wild noise at the amp when it's running. I'll check my wiring and see if I can resolve that first...

**Comment by musiq021** [comment]
I re-assembled the circuit with more care on a higher quality breadboard and the noise coming from the amplifier during the SD card activity is gone.

- The mic -> ram -> amp sketch is working.
- The SD test sketch is working.

Still crashing when calling my startRec function

**Comment by pschatzmann** [comment]
I think you messed up badly: don't mix different sd libraries !

**Comment by musiq021** [comment]
Ah! Ok, changed to..
#include <SD.h>
#include "AudioLibs/AudioSourceSD.h"
Same result though :-\

**Comment by pschatzmann** [comment]
Are you sure that the directory exists? 
Did you get a stacktrace ?
Try to find at which line it is crashing...

**Comment by musiq021** [comment]
The directory exists, but to be sure, I wrote to root dir. No dice. SD.open seems a success. Unfortunately, I don't know if it is just the ESP32-S2 with integrated USB serial or what, but no stacktrace, and I don't see an option to enable it in the IDE. I will try this on an ESP32 Dev board...

**Comment by musiq021** [comment]
Ok, on the Dev module I get:

[I] StreamCopy.h : 82 - buffer_size=1024

Stack smashing protect failure!

abort() was called at PC 0x400e8c0c on core 1


Backtrace: 0x4008378d:0x3ffb20d0 0x40089be1:0x3ffb20f0 0x4008eebd:0x3ffb2110 0x400e8c0c:0x3ffb2190 0x400da7ec:0x3ffb21b0 0x400da7fc:0x3ffb2200 0x400da959:0x3ffb2220 0x400d8e8a:0x3ffb2240 0x400e00f9:0x3ffb2290

ELF file SHA256: 7c158e6161395560

Rebooting...

**Comment by musiq021** [comment]
0x4008378d: panic_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/panic.c:408
0x40089be1: esp_system_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/esp_system.c:137
0x4008eebd: abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/newlib/abort.c:46
0x400e8c0c: find_task_in_twdt_list at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/task_wdt.c:97
0x400da7ec: _GLOBAL__sub_D_dirSounds at /Users/me/Documents/Arduino/echoTaki/et_sd_wav_player/et_sd_wav_player.ino:205
0x400da7fc: __static_initialization_and_destruction_0 at /Users/me/Documents/Arduino/libraries/ebl_ButtonEvent/src/ButtonEvent.cpp:173
0x400da959: ButtonEventClass::loop() at /Users/me/Documents/Arduino/libraries/ebl_ButtonEvent/src/ButtonEvent.cpp:112
0x400d8e8a: __atomic_add at /Users/me/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/esp-2021r2-patch5-8.4.0/xtensa-esp32-elf/include/c++/8.4.0/ext/atomicity.h:53
0x400d8e8a: __atomic_add_dispatch at /Users/me/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/esp-2021r2-patch5-8.4.0/xtensa-esp32-elf/include/c++/8.4.0/ext/atomicity.h:96
0x400d8e8a: std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_add_ref_copy() at /Users/me/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/esp-2021r2-patch5-8.4.0/xtensa-esp32-elf/include/c++/8.4.0/bits/shared_ptr_base.h:139
0x400d8e8a: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count(std::__shared_count<(__gnu_cxx::_Lock_policy)2> const&) at /Users/me/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/esp-2021r2-patch5-8.4.0/xtensa-esp32-elf/include/c++/8.4.0/bits/shared_ptr_base.h:735
0x400d8e8a: std::__shared_ptr::__shared_ptr(std::__shared_ptr const&) at /Users/me/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/esp-2021r2-patch5-8.4.0/xtensa-esp32-elf/include/c++/8.4.0/bits/shared_ptr_base.h:1165
0x400d8e8a: std::shared_ptr::shared_ptr(std::shared_ptr const&) at /Users/me/Library/Arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/esp-2021r2-patch5-8.4.0/xtensa-esp32-elf/include/c++/8.4.0/bits/shared_ptr.h:129
0x400d8e8a: fs::File::File(std::shared_ptr) at /Users/me/Library/Arduino15/packages/esp32/hardware/esp32/2.0.8/libraries/FS/src/FS.h:50
0x400e00f9: app_main at /Users/me/Library/Arduino15/packages/esp32/hardware/esp32/2.0.8/cores/esp32/main.cpp:71

**Comment by pschatzmann** [comment]
So, it is quite obvious: et_sd_wav_player.ino line 205 - I don't know anything about your ButtonEvent, so I can't provide any help!

**Comment by musiq021** [comment]
I commented out all code relating to the button event library. Here is the resultant interpreted backtrace : 

0x4008378d: panic_abort at /Users/me/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/panic.c:408
0x40089be1: esp_system_abort at /Users/me/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/esp_system.c:137
0x4008eebd: abort at /Users/me/Desktop/ESP32/ESP32S2/esp-idf-public/components/newlib/abort.c:46
0x400e7b0c: __stack_chk_fail at /Users/me/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/stack_check.c:36
0x400da009: recStart() at /Users/me/Documents/Arduino/echoTaki/et_system_test_b/et_system_test_b.ino:132
0x400da120: loop() at /Users/me/Documents/Arduino/echoTaki/et_system_test_b/et_system_test_b.ino:101
0x400df7f1: loopTask(void*) at /Users/me/Library/Arduino15/packages/esp32/hardware/esp32/2.0.8/cores/esp32/main.cpp:50

**Comment by pschatzmann** [comment]
Just look at the line in the sketch where the issues starts and follow the calls: Google is your friend.
My first guess would be that you write to some stack data outside of the reserved range!

I don't like if I have to repeat myself, but  https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working still applies!

**Reply by musiq021** [reply]
Thanks Phil, it seems I still have much to learn about C. I'm finally starting to make sense of pointers, but the stack is a mysterious black box to me yet.

You're right, it had nothing to do with your library, the line "sprintf(newFileName, "REC-%06d.wav", rand() % 1000000);" was writing to a 14 character buffer, but I forgot the null terminator :-( . 

I am in awe that you were able to put together such an impressive library. Debugging has been a huge challenge for me. 

Again, I really appreciate you being willing to help out a random stranger noob for the millionth time.

**Comment by pschatzmann** [comment]
Maybe you invest also some time to learn the classes that are available in [Arduino](https://www.arduino.cc/reference/en/). 

It has it's own classes: e.g. String. Arduino is a basically C++, so the corresponding classes will also work most of the time: e.g. std::string. However I agree that it is often more efficient to use the C functionality. 

If you define a variable in a method, it is allocated on the stack and automatically disappears, when the method is left.
